Add firewall section to the networking page

larskarlitski opened the pull request:
First look at a what a firewall module might look like.

It operates on firewalld's default zone and there are no plans to expose zones, as they don't seem to be used very often on servers.

The delete buttons don't work yet, but the UI updates itself when adding and removing services from the default zone with one of:
```
sudo firewallctl zone '' remove service ssh
sudo firewallctl zone '' add service ssh
``` 

![firewall](https://user-images.githubusercontent.com/1221707/32411073-f94cb802-c1d1-11e7-80aa-d9d0226b5243.png)

This will implement the minimal version of the design and iterate in later pull requests.

- [x] show the firewall section on the default page when firewalld is installed
- [x] show the list of services that are currently active
- [x] allow to remove a service
- [x] allow to add one of firewalld's predefined services
- [x] tests
- [x] add doc/guide page
- [ ] feature video for release notes

garrett commented:
Related links:

- design page: https://github.com/cockpit-project/cockpit/wiki/Feature:-Firewall
- issue: https://github.com/cockpit-project/cockpit/issues/1094

q2dg commented:
Ei, there's any still-alpha firewall module available to download for trying it? Thanks!

larskarlitski commented:
First update towards @garrett's design: move the page over to the network page.

Small bug (in addition to all the stuff missing): `cockpit.location.go()` in the breadcrumb doesn't do what it should.

larskarlitski commented:
Quick update: make the breadcrumb work correctly, hook up the switch to starting/stopping `firewalld.service`, and hide the whole section if firewalld is not installed.

larskarlitski commented:
Update with a refactor, because my initial approach of talking to firewalld turned out to make too many D-Bus calls and cause too many re-renders. Removing services is now supported. Also lots of tiny fixes to get it close to the design.

andreasn commented on pkg/networkmanager/firewall.jsx:
```diff
@@ -0,0 +1,167 @@
+        return (
+            <div className="container-fluid page-ct">
+                <ol className="breadcrumb">
+                    <li><a onClick={go_up}>{_("Network")}</a></li>
```
Just a flyby comment that this is called "Networking" in the menu and on the breadcrumb on the other sub-pages.

larskarlitski replied:
Indeed. Thanks for the catch! It'll be fixed with the next push.

larskarlitski commented:
Update with minor fixes pointed out by @martinpitt and @andreasn. Also add a first version of the "Add Services" dialog. It's still missing the search box and some layout fixes to qualify for this PR, so leaving WIP.

Depends on #8794 now.

larskarlitski commented:
Fix a couple of things we discussed on IRC: 

* "N Active Rules" on the main networking page is now clickable (and blue on hover)
* services on the firewall page and in the "Add Services" dialog are now (actually) sorted alphabetically
* clicking anywhere on a row in the "Add Services" dialog toggles its checkbox
* fix alignment and spacing issues

larskarlitski commented:
Another update, which I think brings it to a state for a first round of review (removed WIP):
* add search box in the Add Services Dialog
* skip the tests on some images that don't have firewalld
* fix networking test that used ambiguous selector now, because of the new elements

martinpitt commented on pkg/networkmanager/firewall.jsx:
```diff
@@ -0,0 +1,265 @@
+        super();
+
+        this.state = {
+            selected: new Set()
```
Please initialize `services` here (to `null` so that your spinner still works). At least I prefer the state components to be spelled out here in full, for an easier mental model.

martinpitt commented on pkg/networkmanager/firewall.jsx:
```diff
@@ -0,0 +1,265 @@
+        var enabled = event.target.checked;
+
+        this.setState(oldState => {
+            var selected = new Set(oldState.selected);
```
Whee, functional `setState()` - I've learned something new today :+1: 

Please use `let` in arrow functions for correct scoping.

martinpitt commented on pkg/networkmanager/firewall.jsx:
```diff
@@ -0,0 +1,265 @@
+            enabled: firewall.enabled,
+            services: firewall.services,
+            enabledServices: firewall.enabledServices
+        };
```
It feels redundant to keep a verbatim copy of the global `firewall` state here and in `onFirewallChanged`. Isn't it enough to just call `setState({ firewall })` in `onFirewallChanged`, and initialize it the same way here? Or am I missing something?

martinpitt replied:
I'd still like to discuss the above.

martinpitt commented on pkg/networkmanager/firewall.jsx:
```diff
@@ -0,0 +1,265 @@
+    }
+
+    onAddServices() {
+        var services = [...this.state.services].map(id => this.state.services[id]);
```
Why does this need a copy of state.services? The .map already creates a new object.

larskarlitski replied:
I agree, this is very unfortunate. `this.state.services` is a `Set`, which doesn't have `map` defined.

martinpitt commented on pkg/networkmanager/interfaces.js:
```diff
@@ -2144,6 +2145,11 @@ PageNetworkInterface.prototype = {
         });
 
+        $("#networking-firewall-link, #networking-firewall-summary").on("click", function() {
+            cockpit.jump("/network/firewall");
```
Indeed it is not necessary any more to specify the host in master, but in the interest of backportability, and for working with older cockpit-system in RHEL, please add it here.

martinpitt commented on pkg/networkmanager/firewall.jsx:
```diff
@@ -0,0 +1,265 @@
+        if (!this.state.installed) {
+            return (
+                <EmptyState title={_("Firewall is not available")} icon="fa fa-exclamation-circle">
+                    <p translatable="yes">Please install the <tt>firewalld</tt> package.</p>
```
Can you please update the .pot and make sure that this actually works? I thought the `translate` (aka `translatable`) tags were only considered in .html files. If it doesn't, use `cockpit.format(_("Please ... {0} package"), "firewalld")` here.

martinpitt commented on pkg/networkmanager/firewall.jsx:
```diff
@@ -0,0 +1,265 @@
+
+        var addServiceAction = <button className="btn btn-primary pull-right" onClick={this.onAddServices}>{_("Add Services…")}</button>;
+
+        var services = [...this.state.enabledServices].map(id => this.state.services[id]);
```
Again, why does enabledServices need to be copied?

martinpitt commented on test/verify/check-networking-firewall:
```diff
@@ -0,0 +1,108 @@
+    while machine.execute("systemctl is-active {} || true".format(unit)).strip() != state:
+        time.sleep(0.1)
+        if time.time() - start > timeout:
+            raise Exception("timed out waiting for {} to become {}".format(unit, state))
```
Consider using testlib.py's `wait()` which has the timeout and graceful exception handling etc. builit in. Something like

    wait(lambda: machine.execute("systemctl is-active " + unit).strip() != state, delay=0.2)

(with the default of `tries=60` you then get a 12s timeout)

martinpitt commented on test/verify/check-networking-firewall:
```diff
@@ -0,0 +1,108 @@
+            raise Exception("timed out waiting for {} to become {}".format(unit, state))
+
+
+class TestNetworking(MachineCase):
```
Please rename to `TestFirewall`, to avoid confusion.

These tests need to be blacklisted on at least rhel-7-4 (where networking is in base, and thus don't get this feature). There are also some images (like the Debians) which don't have firewalld, I suggest the test skips them for now.

martinpitt commented on test/verify/check-networking-firewall:
```diff
@@ -0,0 +1,108 @@
+        wait_unit_state(m, "firewalld", "active")
+
+        active_rules = m.execute("firewallctl zone '' list services").split()
+        b.wait_in_text("#networking-firewall-summary", "{} Active Rules".format(len(active_rules)))
```
Corner case: Use "Rule" here instead of "Rules", for the corner case that there is really just one. However, our images should have at least ssh and cockpit, so I guess it's fine.

larskarlitski replied:
Wow, awesome catch!

martinpitt commented on test/verify/check-networking-firewall:
```diff
@@ -0,0 +1,108 @@
+        b.wait_present("#networking-firewall-switch label.active:not(.disabled)")
+        b.wait_in_text("#networking-firewall-switch label.active", "Off")
+        wait_unit_state(m, "firewalld", "inactive")
+        b.wait_in_text("#networking-firewall-summary", "0 Active Rules")
```
This behaviour feels a bit odd. If the firewall is disabled, then "0 active rules" feels like "enabled, but unconfigured". Shouldn't this be hidden completely, as the firewall state is already shown as "off"?

larskarlitski replied:
This is as designed. I agree that we might want something better here. I'll ask @garrett.

martinpitt commented on test/verify/check-networking-firewall:
```diff
@@ -0,0 +1,108 @@
+        b.enter_page("/network/firewall")
+
+        b.click(".breadcrumb li:first a")
+        b.enter_page("/network")
```
I suggest stopping and removing firewalld.service from disk (don't forget systemctl daemon-reload), and verify the expected behaviour of "not installed".

martinpitt commented on test/verify/check-networking-firewall:
```diff
@@ -0,0 +1,108 @@
+
+        self.login_and_go("/network/firewall")
+
+        # Add a service that's probably not enabled by default anywhere
```
... so do the equivalent wait_not_present()?

larskarlitski commented on pkg/networkmanager/firewall-client.es6:
```diff
@@ -0,0 +1,175 @@
+var firewalld_dbus = cockpit.dbus('org.fedoraproject.FirewallD1', { superuser: 'try' });
+
+firewalld_service.addEventListener('changed', function () {
+    var installed = !!firewalld_service.exists;
```
It's set to `null` before the first update. I've been bitten by JavaScript's weird type coercion too many times, so I've left the `!!`.

larskarlitski commented on pkg/networkmanager/firewall-client.es6:
```diff
@@ -0,0 +1,175 @@
+     * that's just confusing.
+     */
+    .then(function () {
+        return Array.prototype.slice.call(arguments);
```
Yeah, it's the most succinct way to do so. I can't use a fat-arrow function here, because that doesn't have the `arguments` object (took me a while to figure out). This looks confusing, because the real culprit here is `cockpit.all()` not returning a single list. I've updated the comment to be more clear.

larskarlitski commented on pkg/networkmanager/firewall-client.es6:
```diff
@@ -0,0 +1,175 @@
+
+                firewall.services[service] = info;
+                return info;
+            });
```
I'm handling those errors where the surrounding function is called.

larskarlitski commented on pkg/networkmanager/firewall-client.es6:
```diff
@@ -0,0 +1,175 @@
+}, (path, iface, signal, args) => {
+    var [ , service, ] = args;
+
+    firewall.enabledServices.delete(service);
```
`firewall.services` needs to contain all services, because it is uses in the "Add Services" dialog. It's used as a cache if you never open the dialog so that we don't need to fetch information on every state update.

larskarlitski commented on pkg/networkmanager/firewall-client.es6:
```diff
@@ -0,0 +1,175 @@
+firewall.addServices = function (services) {
+    return cockpit.all(services.map(s => firewall.addService(s)))
+        .then(function () {
+            return Array.prototype.slice.call(arguments);
```
Yeah, they return promises that resolve when the service is added. I've documented all those functions now so that that is more clear.

larskarlitski commented on pkg/networkmanager/firewall.jsx:
```diff
@@ -0,0 +1,265 @@
+
+var _ = cockpit.gettext;
+
+require("page.css");
```
Good catch! I think I used it for *something* at some point...

larskarlitski commented on pkg/networkmanager/firewall.jsx:
```diff
@@ -0,0 +1,265 @@
+function EmptyState(props) {
+    return (
+        <div className="curtains-ct blank-slate-pf">
+            {props.icon && <div className="blank-slate-pf-icon"><div className={props.icon}></div></div>}
```
Ugh, how did this happen?

larskarlitski commented on pkg/networkmanager/firewall.jsx:
```diff
@@ -0,0 +1,265 @@
+                           columns={["", "", ""]} />
+
+    var tcp = props.service.ports.filter(p => p.protocol.toUpperCase() == 'TCP');
+    var udp = props.service.ports.filter(p => p.protocol.toUpperCase() == 'UDP');
```
Yes, but only very few services make use of that. According to the docs:

> The protocol can be any protocol supported by the system. Please have a look at /etc/protocols for supported protocols.

larskarlitski commented on pkg/networkmanager/firewall.jsx:
```diff
@@ -0,0 +1,265 @@
+                       rowId={props.service.id}
+                       columns={columns}
+                       tabRenderers={tabs}/>;
+}
```
I usually do this when I write components in `pkg/lib`, i.e. when they're supposed to be used in different files from where they are defined. So I haven't added them here either. Let me know if you feel stronger about it than I do.

martinpitt replied:
It's fine for me, just wanted to point it out.

larskarlitski commented:
Thanks very much for the detailed review. Lots of good stuff!

To not spam you too much, I haven't commented "Fixed" on the suggestions that I accepted exactly as you wrote them.

The one thing I haven't fixed is showing the UI read-only when we're not superuser, as that's a bit more involved. How do we usually find this out? Ask polkit directly?

martinpitt commented:
@larskarlitski : We spoke about this in IRC, but for the record: You can query https://www.freedesktop.org/software/polkit/docs/latest/eggdbus-interface-org.freedesktop.PolicyKit1.Authority.html#eggdbus-method-org.freedesktop.PolicyKit1.Authority.CheckAuthorization to see if your process is allowed to authorize. Of course you still need to know whether the user ticked the "use my password for admin tasks", I hope we store that somewhere?

martinpitt commented on pkg/networkmanager/firewall-client.es6:
```diff
@@ -120,6 +121,12 @@ firewalld_dbus.subscribe({
     firewall.dispatchEvent('changed');
 });
 
+cockpit.spawn(['sh', '-c', 'pkcheck --action-id org.fedoraproject.FirewallD1.all --process $$ -u 2>&1'])
```
A bit ugly indeed, but I can't think of a better way. If we encounter this more often in the future, we might instead query/store  the bridge pid as a global variable somewhere.

larskarlitski replied:
I agree. If we start doing this in more places, we probably also want (internal) library functions for that.

martinpitt commented on pkg/networkmanager/firewall.jsx:
```diff
@@ -270,7 +272,12 @@ export class Firewall extends React.Component {
+        var addServiceAction;
+        var classes = "btn btn-primary pull-right";
+        if (this.state.firewall.readonly)
+            addServiceAction = <button className={classes} disabled title={_("You are not authorized to modify the firewall.")}>{_("Add Services…")}</button>;
```
This is using the standard browser tooltips (white on black) which are "wrong". I suggest using `update_privileged()` for that, or if that somehow is impractical, use `.tooltip('fixTitle')` (but that requires jQuery) or our `Tooltip` react component.

larskarlitski replied:
`update_privileged()` also depends on jQuery, which I don't want to pull into `firewall.jsx`. I've used the React component, which unfortunately needed some minor tweaking. I've added the two commits on here as well.

martinpitt commented on pkg/networkmanager/firewall.jsx:
```diff
@@ -65,6 +65,8 @@ function ServiceRow(props) {
             { udp.map(p => p.port).join(', ') }
         </div>,
         <button className="btn btn-danger pficon pficon-delete"
+                title={props.readonly ? _("You are not authorized to modify the firewall.") : null}
```
Also the wrong tooltip. Here it seems mostly irrelevant as an unprivileged user cannot even see the current firewall rules. But there is the theoretical possibility that an admin changes the polkit rules to allow that to users, but not modification.

larskarlitski replied:
I can see the firewall rules as an unprivileged user on fedora. Fixed as above.

larskarlitski commented:
> This works well enough, thanks for that! The "On/Off" button on the /network/firewall page still needs the same "disable" treatment as "Add services..". I can't actually operate it in unprivileged mode - it stays at "off" even though the firewall is really on.

As discussed on IRC: I can operate the button even in unprivileged mode. I guess there's something wrong with `superuser: "try"`, but that's unrelated to this pull request.

> I don't see the firewall section on /network at all in unprivileged mode. I think this is right - the /firewall page is next to useless then anyway.

I do, but I've left it the way it is for the reason above.

I've added two more commits to fix our `Tooltip` component, but left the extra commit for the unprivileged stuff so that it's easier for you to review. I'll squash it into the main commit after the review.

martinpitt commented:
> I can operate the button even in unprivileged mode

Actually now I see that too, not sure what was different. However, does that actually work for you? What happens here is that the firewall is shown as "off", then I can click it to "on" (but does not actually start firewalld.service), and then it moves to the disabled state. From then on it stays disabled.

Another issue is that in privileged mode the On/Off switch is a bit confusing. It takes about two seconds after clicking to actually take effect, but during that time the button stays enabled and does not change state. This feels like "I mis-clicked", and when clicking again it interferes with the first click.

Compare this with the "Automatic Updates" on/off switch on software updates. This immediately switches state, and gets disabled until the action takes place. This was the result of some involved discussions with our designers, so I think we should have the same behaviour here.

Both are minor issues, so I would be happy if these become follow-ups if you prefer. The rest looks good, great job with figuring out the tooltip issues! I confirm that the changes work well, also on the kdump/playground pages.

larskarlitski commented:
I've changed the switch on the firewall page to behave in the same way as the other one: it is disabled while the state change is pending.

The tests had two issues: one was a logic mixup (`!=` vs `==`) and the other one that we need to wait for the Add Services button to become enabled before we can click it. The latter happened only after I added the read-only mode, which is the default and the button can be disabled for a few ms.

Also folded the read-only mode commit into the main one.

Sorry for the delay.

larskarlitski commented:
Still some test failures.

martinpitt commented:
Hmm, this collided with #9027. Maybe this was too aggressive and we should revert the "indent" stuff..

martinpitt commented:
@larskarlitski : Force-pushed with running `npm run eslint:fix` to get the indentation right, and with capitalizing the commit messages.

larskarlitski commented:
Ah, I rebased and fixed a conflict,  it didn’t build again, because I had to run. Thanks!

apollo13 commented on pkg/networkmanager/firewall-client.es6:
```diff
@@ -0,0 +1,198 @@
+    firewall.dispatchEvent('changed');
+});
+
+cockpit.spawn(['sh', '-c', 'pkcheck --action-id org.fedoraproject.FirewallD1.all --process $$ -u 2>&1'])
```
Maybe replace `-u` with `--allow-user-interaction` -- short options are nice for users to type; but not really in code imo.

martinpitt commented:
eslint spotted a bug and has another complaint:
```
/build/cockpit/pkg/networkmanager/firewall.jsx
  179:36  error  Unknown property 'for' found, use 'htmlFor' instead  react/no-unknown-property
  253:64  error  Arrow function should not return assignment          no-return-assign
```

martinpitt commented on test/verify/check-networking-firewall:
```diff
@@ -0,0 +1,117 @@
+    def active_state(unit):
+        # don't use `systemctl is-active` here because of https://bugzilla.redhat.com/show_bug.cgi?id=1073481
+        # outputs "ActiveState=state\n"
+        line = machine.execute("systemctl show -p ActiveState {}".format(unit))
```
Note: `--value` does not yet exist in current CentOS 7.4. Once that moves to 7.5, we can use it.

larskarlitski replied:
Ah I didn't know about that. Thank you!

martinpitt commented on test/verify/check-networking-firewall:
```diff
@@ -0,0 +1,117 @@
+    wait(lambda: active_state(unit) == state, delay=0.2)
+
+
+@skipImage("firewalld not installed", "rhel-7-5", "centos-7", "fedora-atomic", "rhel-atomic", "debian-stable", "debian-testing", "ubuntu-stable", "ubuntu-1604")
```
Follow-up: We should definitively install it on rhel-7-5 and centos-7 at least. So on rhel-8 it's already present apparently, good!

larskarlitski commented on doc/guide/feature-firewall.xml:
```diff
@@ -0,0 +1,27 @@
+    services in the default zone.</para>
+
+  <para>To perform similar tasks from the command line, use
+    <ulink url="http://www.firewalld.org/documentation/man-pages/firewall-cmd.html">firewalld-cmd</ulink>:
```
I had tried `https://firewalld.org` and that didn't work. It does with the `www` though. Good catch!

larskarlitski commented on doc/guide/feature-firewall.xml:
```diff
@@ -0,0 +1,27 @@
+<programlisting>
+$ <command>sudo firewall-cmd --list-services</command>
+dhcpv6-client samba-client mdns ssh cockpit
+</programlisting>
```
Yeah good point.

larskarlitski commented on pkg/networkmanager/firewall-client.es6:
```diff
@@ -0,0 +1,198 @@
+     */
+            .then(function () {
+                return Array.prototype.slice.call(arguments);
+            });
```
There's a comment above these lines which explains why this is: `cockpit.all` returns results as individual arguments, so we need access to the `arguments` object, which doesn't exist in arrow functions.

Also, eslint gets the indentation wrong here. Do you know why?

martinpitt replied:
I don't, I'm afraid. You mean the alignment of the `});`, right? If some rule gets it wrong, feel free to disable it for the time being.

larskarlitski commented on pkg/networkmanager/firewall-client.es6:
```diff
@@ -0,0 +1,198 @@
+
+    if (!firewall.enabled) {
+        firewall.services = {};
+        firewall.enabledServices = new Set();
```
Indeed! Very nice catch, thanks.

larskarlitski commented on pkg/networkmanager/firewall.jsx:
```diff
@@ -0,0 +1,342 @@
+
+    componentDidMount() {
+        firewall.getAvailableServices()
+                .then(services => this.setState({ services: services }));
```
Hm, this is a normal `setState()`, not functional?!

martinpitt replied:
Err yes, clearly I grossly misread that.. :-/

larskarlitski commented on pkg/networkmanager/firewall.jsx:
```diff
@@ -0,0 +1,342 @@
+    }
+
+    onAddServices() {
+        var services = [...this.state.firewall.services].map(id => this.state.firewall.services[id]);
```
You did, but no worries ;) `map()` only exists on Arrays, not Objects.

larskarlitski commented on pkg/networkmanager/firewall.jsx:
```diff
@@ -0,0 +1,342 @@
+        if (!this.state.firewall.installed) {
+            return (
+                <EmptyState title={_("Firewall is not available")} icon="fa fa-exclamation-circle">
+                    <p>{cockpit.format(_("Please install the {0} package"), "firewalld")}</p>
```
Indeed!

larskarlitski commented on pkg/networkmanager/interfaces.js:
```diff
@@ -2231,6 +2237,33 @@ PageNetworkInterface.prototype = {
+            $('#networking-firewall-switch').onoff('value', firewall.enabled);
+
+            var n = firewall.enabledServices.size;
+            var summary = cockpit.ngettext('1 Active Rule', n + ' Active Rules', n);
```
Oooooooops

larskarlitski commented:
Thanks very much for the review! The ones I've left uncommented are fixed as well. Didn't want to spam you too much ;)

martinpitt commented:
Oops, the [test still fails](https://fedorapeople.org/groups/cockpit/logs/pull-8032-20180426-192020-a48e7317-verify-fedora-28/log.html#170): The [screenshot](https://fedorapeople.org/groups/cockpit/logs/pull-8032-20180426-193434-a48e7317-verify-fedora-i386/TestFirewall-testNetworkingPage-fedora-i386-127.0.0.2-2601-FAIL.png) shows that it's literally displayed `$0 Active Rules`.

martinpitt commented on pkg/networkmanager/interfaces.js:
```diff
@@ -2231,6 +2237,33 @@ PageNetworkInterface.prototype = {
+            $('#networking-firewall-switch').onoff('value', firewall.enabled);
+
+            var n = firewall.enabledServices.size;
+            var summary = cockpit.ngettext('$0 Active Rule', '$0 Active Rules', n);
```
Sorry, this needs to be wrapped in `cockpit.format(cockpit.ngettext(...), n)` to actually resolve the `$0`.

larskarlitski commented:
Added `cockpit.format()` (and a hack, because that cannot handle `0` as an argument) and reordered the location with the weird indent to make that more readable.

martinpitt commented:
Corresponding semaphore run: https://semaphoreci.com/cockpit/cockpit/branches/pull-request-8032/builds/20 (wasn't added as test here for some reason)

openface commented:
Just coming into this project and am really impressed.  In particular, am glad to see support for firewalld. 

Are there any plans to allow specific firewall configuration per network interface?
